### PR TITLE
storage/inmem: Avoid unnecessary read operation

### PR DESCRIPTION
**Pull request by ashutosh-narkar**
While writing data to the in-memory store via truncate op,
OPA reads data at the current path before adding
new data to the store to ensure the path exists.

A bundle that contains data files at non-root locations will
trigger a read on the store for each file and hence for a large
bundle this can cause an increase in the bundle activation time
and also resource usage.

This change attempts to avoid multiple read ops by merging all the data
in the bundle and performing a single write on the store.

### CODE REVIEW

File: storage/inmem/inmem.go [hunk 1]
```diff
@@ -121,20 +121,19 @@ func (db *store) Truncate(ctx context.Context, txn storage.Transaction, _ storag
 					return err
 				}
 
-				_, err := underlying.Read(update.Path[:len(update.Path)-1])
+				err = underlying.Write(storage.AddOp, update.Path, obj)
```
Comment (srenatus) [inline storage/inmem/inmem.go:124]: Hmm this reads funny. If I understand this correctly, we're attempting the write, and if the path isn't there, we makedir it, and write again. Is MakeDir costly? We could always do it first and write afterwards 🤔

Comment (ashutosh-narkar) [reply storage/inmem/inmem.go:124]: So the idea is that the first `Write` call will fail as the storage won't exist but the next ones should work w/o needing to read the path first.

**Review by srenatus — APPROVED** [review]

File: storage/inmem/inmem.go (hunk 1 above)
Comment (ashutosh-narkar) [reply storage/inmem/inmem.go:124]: I'm trying an alternate approach where we don't rely on the error status from the Write call. Let's see how that goes. Thanks for the review!

**Review by srenatus — APPROVED** [review]
Code looks good, a couple of nitpicks and questions inline. 

There are few things I've thought about outside of the code:
1. Is it correct that most of the behaviour is already covered by tests -- so we change the code, but the existing tests still pass, so we're OK
2. We do this for performance, and it's hard to assert that the new code is performing better than the old. Perhaps you could add a note to the commit message on what you've done to verify that it's better now?

File: storage/inmem/inmem_test.go [hunk 2]
```diff
@@ -505,7 +505,7 @@ func TestTruncateBadRootWrite(t *testing.T) {
 		t.Fatal("Expected truncate error but got nil")
 	}
 
-	expected := "storage_invalid_patch_error: root must be object"
+	expected := "root value must be object"
```
[the remark this answers is not in the document]
Comment (ashutosh-narkar) [reply storage/inmem/inmem_test.go]: Sure. Updated the error message to have the prefix.

File: storage/inmem/inmem.go [hunk 3]
```diff
@@ -168,6 +150,25 @@ func (db *store) Truncate(ctx context.Context, txn storage.Transaction, _ storag
...
+		}
+
+		if len(newPath) > 0 {
+			if err := storage.MakeDir(ctx, db, txn, newPath[:len(newPath)-1]); err != nil {
+				return err
+			}
+		}
+
+		err = underlying.Write(storage.AddOp, newPath, mergedData[k])
+		if err != nil {
```
[the remark this answers is not in the document]
Comment (ashutosh-narkar) [reply storage/inmem/inmem.go]: Updated

File: storage/inmem/inmem.go [hunk 4]
```diff
@@ -114,52 +119,29 @@ func (db *store) Truncate(ctx context.Context, txn storage.Transaction, _ storag
...
 
-					if err := storage.MakeDir(ctx, db, txn, newPath[:len(newPath)-1]); err != nil {
-						return err
-					}
-
-					err = underlying.Write(storage.AddOp, newPath, val[k])
-					if err != nil {
-						return err
-					}
+				merged, ok := merge.InterfaceMaps(mergedData, obj)
```
[the remark this answers is not in the document]
Comment (ashutosh-narkar) [reply storage/inmem/inmem.go]: We can. Updated.

File: storage/inmem/inmem.go [hunk 5]
```diff
@@ -114,52 +119,29 @@ func (db *store) Truncate(ctx context.Context, txn storage.Transaction, _ storag
...
+				if !ok {
+					return fmt.Errorf("failed to insert data file from path %s", filepath.Join(key...))
```
[the remark this answers is not in the document]
Comment (ashutosh-narkar) [reply storage/inmem/inmem.go:143]: Added a test case.

**Review by srenatus — DISMISSED** [review]
Thank you

**Pull request merged**